apps/emqx_rule_engine/test/emqx_rule_engine_api_rule_apply_SUITE.erl

```diff
@@ -413,6 +397,246 @@ t_apply_rule_test_batch_separation_stop_after_render(_Config) ->
+                        <<"trace_tag">> := <<"ERROR">>
+                    },
+                <<"msg">> := SendErrorMsg,
+                <<"time">> := _
+            },
+            ReasonEntryJSON
+        )
+    end.
+
+meck_test_connector_recoverable_errors(Reason) ->
+    MeckOpts = [passthrough, no_link, no_history, non_strict],
+    catch meck:new(emqx_rule_engine_test_connector, MeckOpts),
+    meck:expect(
+        emqx_rule_engine_test_connector,
+        on_query,
+        3,
+        {error, {recoverable_error, Reason}}
+    ),
+    meck:expect(
+        emqx_rule_engine_test_connector,
+        on_batch_query,
+        3,
+        {error, {recoverable_error, Reason}}
+    ),
+    meck:expect(
+        emqx_rule_engine_test_connector,
+        on_query_async,
+        4,
+        {error, {recoverable_error, Reason}}
+    ),
+    meck:expect(
+        emqx_rule_engine_test_connector,
+        on_batch_query_async,
+        4,
+        {error, {recoverable_error, Reason}}
+    ).
+
+find_lines_with(Data, InLineText) ->
+    % Split the binary data into lines
+    Lines = re:split(Data, "\n", [{return, binary}]),
+
+    % Use a list comprehension to filter lines containing 'Reason'
+    [Line || Line <- Lines, re:run(Line, InLineText, [multiline, {capture, none}]) =/= nomatch].
+
+do_apply_rule_test_format_action_failed_test(BatchSize, CheckLastTraceEntryFun) ->
+    meck_in_test_connector(),
+    {ok, _} = emqx_connector:create(rule_engine_test, ?FUNCTION_NAME, #{}),
+    Name = atom_to_binary(?FUNCTION_NAME),
+    ActionConf =
+        #{
+            <<"connector">> => Name,
+            <<"parameters">> => #{<<"values">> => #{}},
+            <<"resource_opts">> => #{
+                <<"batch_size">> => BatchSize,
+                <<"batch_time">> => 10,
+                <<"request_ttl">> => 200
+            }
+        },
+    {ok, _} = emqx_bridge_v2:create(
+        rule_engine_test,
+        ?FUNCTION_NAME,
+        ActionConf
+    ),
+    SQL = <<"SELECT payload.is_stop_after_render as stop_after_render FROM \"", Name/binary, "\"">>,
+    {ok, RuleID} = create_rule_with_action(
+        rule_engine_test,
+        ?FUNCTION_NAME,
+        SQL
+    ),
+    create_trace(Name, ruleid, RuleID),
+    Now = erlang:system_time(second) - 10,
+    %% Stop
+    ParmsNoStopAfterRender = apply_rule_parms(false, Name),
+    {ok, _} = call_apply_rule_api(RuleID, ParmsNoStopAfterRender),
+    %% Just check that the log file is created as expected
+    ?retry(
+        _Interval0 = 200,
+        _NAttempts0 = 100,
+        begin
+            Bin = read_rule_trace_file(Name, ruleid, Now),
+            CheckLastTraceEntryFun(Bin)
+        end
+    ),
+    %% Cleanup
+    ok = emqx_trace:delete(Name),
+    ok = emqx_rule_engine:delete_rule(RuleID),
+    ok = emqx_bridge_v2:remove(rule_engine_test, ?FUNCTION_NAME),
+    ok = emqx_connector:remove(rule_engine_test, ?FUNCTION_NAME),
+    [_, _, _] = meck:unload(),
```

Comment:
Nit: these could be placed in a `on_exit` call that is close to where those "resources" are created to ensure they'll be cleaned up by the [janitor](https://github.com/emqx/emqx/blob/e19222fc0d079f95c17480699aee69c5d5f1d4cc/apps/emqx_rule_engine/test/emqx_rule_engine_api_rule_apply_SUITE.erl#L66) even if the assertions fail.

Reply:
Fixed: [d5324e2](https://github.com/emqx/emqx/pull/12981/commits/d5324e295f4bbe9102901c922a64186e45f73fe1)